pyro/poutine/uncondition_messenger.py

```diff
@@ -0,0 +1,22 @@
+from __future__ import absolute_import, division, print_function
+
+from .messenger import Messenger
+
+
+class UnconditionMessenger(Messenger):
+    """
+    Messenger to force the value of observed nodes to be sampled from their
+    distribution, ignoring observations.
+    """
+    def __init__(self):
+        super(UnconditionMessenger, self).__init__()
+
+    def _pyro_sample(self, msg):
+        """
+        :param msg: current message at a trace site.
+
+        Samples value from distribution, irrespective of whether or not the
+        node has an observed value.
+        """
+        msg["value"] = msg["fn"](*msg["args"], **msg["kwargs"])
```

Comment:
The reason I didn't do it in this way before was so that I could use `model_trace.observation_nodes` to iterate over the "observations" in `CSIS._get_matched_trace`. Since I don't think this will work with your suggestion, would it be reasonable to add another flag like:
```
if msg["is_observed"]:
    msg["is_observed"] = False
    msg["value"] = None
    msg["done"] = False
    msg["was_observed"] = True
```
I just want to check this is sufficiently Pyro-ey and won't mess up any other parts.

Reply:
Ah, I see - adding that new flag is fine, except instead of using `msg["was_observed"]` use `msg["infer"]["was_observed"]`

Reply:
Great, will do